http-server/src/main/java/io/micronaut/http/server/RouteExecutor.java

```diff
@@ -433,13 +428,17 @@ public Runnable instrument(Runnable command) {
     }
 
     private boolean isSingle(RouteInfo<?> finalRoute, Class<?> bodyClass) {
-        return finalRoute.isSpecifiedSingle() || (finalRoute.isSingleResult() &&
-            (finalRoute.isAsync() || finalRoute.isSuspended() || Publishers.isSingle(bodyClass)));
+        return finalRoute.isSpecifiedSingle()
+            || (finalRoute.isSingleResult() && (finalRoute.isAsync() || finalRoute.isSuspended()))
+            || Publishers.isSingle(bodyClass);
     }
 
-    private ExecutionFlow<MutableHttpResponse<?>> fromImperativeExecute(PropagatedContext propagatedContext, HttpRequest<?> request, RouteInfo<?> routeInfo, Object body) {
+    private ExecutionFlow<MutableHttpResponse<?>> fromImperativeExecute(PropagatedContext propagatedContext,
+                                                                        HttpRequest<?> request,
+                                                                        RouteInfo<?> routeInfo,
+                                                                        @Nullable Object body) {
         // performance optimization: check for common body types
-        boolean shortCircuit = body instanceof String || body instanceof byte[];
+        boolean shortCircuit = body instanceof String || body instanceof byte[] || body == null;
```

Comment:
```suggestion
        boolean shortCircuit = body instanceof String || body instanceof byte[];
```

This change is not needed, the shortCircuit is specifically for type pollution which null isn't affected by